packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -895,6 +1080,687 @@ void Add(
   }
 }
 
+//Given array containing entries per row, produce rowptrs
+//Used by all add kernels
+template<typename RowPtrType>
+struct PrefixSumFunctor
+{
+  PrefixSumFunctor(RowPtrType& Crowcounts_, RowPtrType& Crowptrs_) : Crowcounts(Crowcounts_), Crowptrs(Crowptrs_) {}
+  //note: call this functor over the full length of Crowptrs (nrows+1)
+  KOKKOS_INLINE_FUNCTION void operator()(const size_t i, size_t& update, const bool finalPass) const
+  {
+    // exclusive scan
+    if(finalPass)
+    {
+      Crowptrs(i) = update;
+    }
+    update += Crowcounts(i);
+    if(finalPass)
+    {
+      if(i == Crowcounts.dimension_0() - 1)
+      {
+        Crowptrs(i + 1) = update;
+      }
+    }
+  }
+  RowPtrType& Crowcounts;
+  RowPtrType& Crowptrs;
+};
+
+/****************************************************
+* Fast local thread-parallel addition using Kokkos *
+ * Adds two Kokkos local CRS matrices               *
+ * Two main code paths: A,B sorted or nonsorted     *
+ ****************************************************/
+namespace AddSortedFunctors
+{
+
+template<typename Ordinal, typename RowPtrs, typename ColInds>
+struct EntryCountingFunctor
+{
+  EntryCountingFunctor(const RowPtrs& Arowptrs_, const ColInds& Acolinds_, const RowPtrs& Browptrs_, const ColInds& Bcolinds_, RowPtrs& Crowcounts_) :
+    Arowptrs(Arowptrs_), Acolinds(Acolinds_),
+    Browptrs(Browptrs_), Bcolinds(Bcolinds_),
+    Crowcounts(Crowcounts_) {}
+  KOKKOS_INLINE_FUNCTION void operator()(const size_t i) const
+  {
+    //count the union of nonzeros in Arow and Brow
+    Ordinal numEntries = 0;
+    Ordinal ai = 0;
+    Ordinal bi = 0;
+    auto Arowstart = Arowptrs(i);
+    auto Arowlen = Arowptrs(i + 1) - Arowstart;
+    auto Browstart = Browptrs(i);
+    auto Browlen = Browptrs(i + 1) - Browstart;
+    while(ai < Arowlen && bi < Browlen)
+    {
+      //have an entry in C's row
+      numEntries++;
+      auto Acol = Acolinds(Arowstart + ai);
+      auto Bcol = Bcolinds(Browstart + bi);
+      if(Acol <= Bcol)
+        ai++;
+      if(Acol >= Bcol)
+        bi++;
+    }
+    numEntries += Arowlen - ai;
+    numEntries += Browlen - bi;
+    Crowcounts(i) = numEntries;
+  }
+  const RowPtrs& Arowptrs;
```

Comment:
See above note on Kokkos::View needing to be stored by value in functors, not by reference.